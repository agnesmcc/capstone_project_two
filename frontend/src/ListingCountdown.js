import { useState, useEffect } from 'react';
import { intlFormatDistance } from 'date-fns';

/**
 * Displays a countdown timer for a listing, updating every second.
 *
 * @param {Object} props
 * @param {string} props.end_datetime - The end datetime of the listing.
 * @param {Function} props.updateListing - A function to be called when the
 *   auction ends.
 * @returns {JSX.Element}
 */
const ListingCountdown = ({ end_datetime, updateListing }) => {
  const [timeRemaining, setTimeRemaining] = useState(null);

  useEffect(() => {
    const calculateTimeRemaining = () => {
      if (end_datetime) {
        const endTime = new Date(end_datetime);
        const currentTime = new Date();
        const timeRemaining = endTime.getTime() - currentTime.getTime();

        if (timeRemaining > 0) {
          const formattedTimeRemaining = intlFormatDistance(endTime, currentTime);
          setTimeRemaining(formattedTimeRemaining);
        } else {
          setTimeRemaining('Ended');
          updateListing();
        }
      }
    };

    calculateTimeRemaining();

    const intervalId = setInterval(calculateTimeRemaining, 1000);

    return () => clearInterval(intervalId);
  }, [end_datetime, updateListing]);

  return (
    <div>
      {timeRemaining && timeRemaining !== 'Ended' ? 
        <div>Auction ends {timeRemaining}</div>
        : <div>Auction has ended</div>}
    </div>
  );
};

export default ListingCountdown;