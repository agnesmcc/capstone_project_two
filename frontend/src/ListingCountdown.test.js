import { render } from '@testing-library/react';
import ListingCountdown from './ListingCountdown';

test('it renders', () => {
  const { container } = render(
    <ListingCountdown end_datetime="2023-07-01T00:00:00" updateListing={() => {}} />
  );
  expect(container).toMatchSnapshot();
});
