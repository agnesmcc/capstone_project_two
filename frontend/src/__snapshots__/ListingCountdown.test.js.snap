// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div>
  <div>
    <div>
      Auction has ended
    </div>
  </div>
</div>
`;
